src/input.rs

```diff
@@ -0,0 +1,182 @@
+    pbr::StandardMaterial,
+    render::mesh::Mesh,
+    transform::components::Transform,
+};
+
+use crate::{
+    chunk::{self, Chunk, CHUNK_SIZE},
+    chunk_manager::ChunkManager,
+    mesher::ChunkMesh,
+    physics::ColliderUpdateEvent,
+    raycaster::{BlockSelection, HighlightCube},
+    world::add_chunk_objects,
+};
+#[derive(Event)]
+pub struct BlockUpdateEvent {
+    pub position: Vec3,
+    pub block: u8,
+}
+#[derive(Event)]
+pub struct ChunkMeshUpdateEvent {
+    pub position: Vec3,
+}
+
+#[derive(Resource)]
+pub struct LastPlayerPosition(pub Vec3);
+
+pub fn handle_mouse_events(
+    mut block_update_events: EventWriter<BlockUpdateEvent>,
+    mut mouse_events: EventReader<MouseButtonInput>,
+    block_selection: Res<BlockSelection>,
+) {
+    if block_selection.normal.is_none() || block_selection.position.is_none() {
+        return;
+    }
+
+    let position = block_selection.position.unwrap();
+    let normal = block_selection.normal.unwrap();
+
+    for event in mouse_events.read() {
+        if event.button == MouseButton::Left && event.state.is_pressed() {
+            block_update_events.send(BlockUpdateEvent { position, block: 0 });
+        } else if event.button == MouseButton::Right && event.state.is_pressed() {
+            block_update_events.send(BlockUpdateEvent {
+                position: position + normal,
+                block: 3,
+            });
+        }
+    }
+}
+
+pub fn handle_keyboard_events(
+    mut keyboard_events: EventReader<KeyboardInput>,
+    camera_query: Query<&Transform, With<HighlightCube>>,
+    mut collider_events: EventWriter<ColliderUpdateEvent>,
+) {
+    for event in keyboard_events.read() {
+        if event.state.is_pressed() {
+            match event.key_code {
+                bevy::input::keyboard::KeyCode::Escape => std::process::exit(0),
+                bevy::input::keyboard::KeyCode::KeyC => {
+                    let controller_transform = camera_query.single();
+                    println!("Handling event: {:?}", controller_transform.translation);
+                    collider_events.send(ColliderUpdateEvent {
+                        position: controller_transform.translation.into(),
+                    });
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+pub fn handle_block_update_events(
+    mut chunk_manager: ResMut<ChunkManager>,
+    mut block_update_events: EventReader<BlockUpdateEvent>,
+    mut chunk_mesh_update_events: EventWriter<ChunkMeshUpdateEvent>,
+) {
+    for event in block_update_events.read() {
+        set_block(event.position, event.block, chunk_manager.as_mut());
+        chunk_mesh_update_events.send(ChunkMeshUpdateEvent {
+            position: event.position / CHUNK_SIZE as f32,
+        });
+    }
+}
+
+pub fn handle_chunk_mesh_update_events(
+    mut commands: Commands,
+    asset_server: Res<AssetServer>,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
```

Comment:
Merge into one collective Resource and clone the handles in order to improve performance and reduce asset count.